### PR TITLE
Add episode 38

### PR DESCRIPTION
**Pull request by colebow**

### CODE REVIEW

File: _episodes/38.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  desc: "Details about features and fixes in the new releases"
+  time:
+---
+
+## Guests
+
+In this episode we have the pleasure to chat with a couple familiar faces who
+have been hard at work on the features we're discussing today:
+
+* [Kasia Findeisen](https://github.com/kasiafi), Trino Contributor
```
Comment (mosabua) [inline _episodes/38.md:19]: maintainer

File: _episodes/38.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+* [Martin Traverso](https://twitter.com/mtraverso), Trino Co-Creator
```
Comment (mosabua) [inline _episodes/38.md:20]: co-creator and maintaine

File: _episodes/38.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+## Demos of the episode: The many ways you can leverage PTFs
+
+### Example 1: Using connector-specific features with query pass-through
+
+Trino always tries to implement the ANSI specification for SQL, and is pretty
+much committed to being compliant with it. However, Trino hasn't yet
+implemented every last function and clause from the ANSI SQL specification, so
+some databases may have features that Trino is yet to get to, and if you want
+to use those features, now you can, no need to wait for Trino to get there!
```
Comment (kasiafi) [inline _episodes/38.md:49]: I wouldn't focus on that so much. Trino is ahead of many databases in standard SQL support. With query pass-through, we will typically address the case when the underlying database has some native non-standard syntax, which cannot be rewritten to standard SQL, or which can be rewritten to standard SQL, but a PTF is used  as a contemporary enabler if someone wants to start using Trino right away, and take more time to adjust their queries. Another use case would be pushdown for performance.

File: _episodes/38.md [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+wanted to, you could use the `query` function to then invoke a PTF in Oracle,
+including any of the hypothetical examples we went into above! PTFs inside of
+PTFs!
+
+## Question of the episode: Where are we at, and what's coming next?
+
+Right now, there's a few things on the radar for how to move forward with PTFs.
+The first and more simple task at hand is expanding the `query` function to
+other connectors. We started with the JDBC connectors, but we have also landed
+it for ElasticSearch and BigQuery, and there may still be more yet to come.
```
Comment (kasiafi) [inline _episodes/38.md:447]: We haven't got the `query` function for BigQuery yet.
Also, the function for Elasticsearch is named `raw_query`.

File: _episodes/38.md [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+
+On a broader scope, the reason this was the first PTF that was implemented is
+because Trino doesn't have to do anything to make it work. The next big step in
+powering PTFs up is to create an operator and make the engine aware of them, so
+that the engine can handle and process PTFs itself, which will open the door to
+the wide array of possibilities we explored earlier.
+
+And finally, once that's done, we plan on empowewring you, the Trino community,
+to go out and actually *make* some polymorphic table functions. Once the
```
Comment (kasiafi) [inline _episodes/38.md:456]: They can add them already. The limitations are: no table or descriptor arguments, and mandatory execution by the connector.

File: _episodes/38.md [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+    Output_pivot_columns DESCRIPTOR,
+    Input_pivot_columns1 DESCRIPTOR,
+    Input_pivot_columns2 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns3 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns4 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns5 DESCRIPTOR DEFAULT NULL
+)
+RETURNS TABLE
+DETERMINISTIC
+READS SQL DATA
```
Comment (kasiafi) [inline _episodes/38.md:245]: Please remove these qualifiers. We don't use them, and we don't determine these properties for table functions in Trino currently.

File: _episodes/38.md [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+#### What can you do with table functions?
+
+So... even standard table functions are cool. They have to return a pre-defined
+schema, which limits their flexibility, but they still have some interesting
+uses as means of shortening queries or performing multiple operations at once.
+If you frequently find yourself selecting from the same table with a `WHERE`
+clause checking equality to a specific column but with a different value each
+time, you could define a table function which takes that value as a parameter
+and allows you to skip all the copying and pasting just for the sake of one line
+changing. You could take an extremely lengthy sub-query with multiple joins and
```
Comment (kasiafi) [inline _episodes/38.md:90]: I'm not sure that this is the kind of usage of PTF we want to suggest.
Such a function would likely be far less efficient that a query with multiple joins executed directly in Trino. And it might be a big challenge to write. In general, if something can be expressed in SQL, it is better to write a normal query, and let Trino optimize and execute it. (However, if someone have got a very specific use case, which they want to optimize in a specific way, writing a PTF might be a good idea)

We're planning to optimize the PTF operator when we have it, to make it a competitive alternative to other operators, but it won't happen soon.

File: _episodes/38.md [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+
+| ID  | Name    | accttype | acctvalue |
+|-----|---------|----------|-----------|
+| 123 | Alice   | external | 20000     |
+| 123 | Alice   | internal | 350       |
+| 234 | Bob     | external | 25000     |
+| 234 | Bob     | internal | 120       |
+
+You can see the `PASS THROUGH` part of the signature working when you select
+D.id and D.name.
```
Comment (kasiafi) [inline _episodes/38.md:305]: With `PASS THROUGH`, all columns of My.Data should be available in the output. Here, the input pivot columns are missing. I noticed the same in the original example in the Technical Report.
I think we can remove the `PASS THROUGH` clause from this example, together with the explanation, and explain the `PASS THROUGH` / `NO PASS THROUGH` in the next example `ExecR`.

File: _episodes/38.md (hunk 7 above)
Comment (colebow) [reply _episodes/38.md:90]: I'd like to keep the example as a really basic one - do you think it's reasonable to leave the example in as a sample of what _can_ be done, but not what _should_ be done?

Comment (colebow) [reply _episodes/38.md:90]: I also didn't realize when commenting this morning - this is the section describing an ordinary (non-polymorphic) table function, which I imagine is less of a concern in terms of performance?

File: _episodes/38.md (hunk 8 above)
Comment (colebow) [reply _episodes/38.md:305]: I don't think they're missing? It's just that the query is only selecting a few specific columns, and isn't doing `SELECT *`

File: _episodes/38.md (hunk 7 above)
Comment (kasiafi) [reply _episodes/38.md:90]: >I also didn't realize when commenting this morning - this is the section describing an ordinary (non-polymorphic) table function, which I imagine is less of a concern in terms of performance?

Polymorphic functions are more expressive than monomorphic, but both types are executed in the same way, with the same performance challenges. Also, from the user's perspective, they are invoked in the same way.

File: _episodes/38.md (hunk 8 above)
Comment (kasiafi) [reply _episodes/38.md]: Ah, right.

File: _episodes/38.md [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+
+## Guests
+
+In this episode we have the pleasure to chat with a couple familiar faces who
+have been hard at work building and understanding the features we're talking
+about today:
+
+### [Kasia Findeisen](https://github.com/kasiafi), Trino Maintainer
+
+[TODO - Kasia]:<>
```
Comment (colebow) [inline _episodes/38.md:22]: @kasiafi - do you want to write a blurb here about how you're going to introduce yourself, how you've worked on PTF, etc? Only needs to be 1-3 sentences.

File: _episodes/38.md [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+
+### [Martin Traverso](https://twitter.com/mtraverso), Trino Co-Creator & Maintainer
+
+[TODO - Martin]:<>
```
Comment (colebow) [inline _episodes/38.md:26]: @martint if there's anything you want to add here, please feel free to suggest it.

File: _episodes/38.md [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+* Improved UNNEST performance.
+* Support for Databricks runtime 10.4 LTS in Delta Lake connector.
+
+[Trino 391](https://trino.io/docs/current/release/release-391.html)
+
+* Support for AWS Athena partition projection.
+* Faster writing of Parquet data in Iceberg and Delta Lake.
+* Support for reading BigQuery external tables.
+* Support for table and column comments in BigQuery.
+
```
Comment (mosabua) [inline _episodes/38.md:63]:
```suggestion
Additional highlights and notes according to Manfred:

* [Java 17 arrived as required runtime in 390]({% post_url 2022-07-14-trino-updates-to-java-17 %}).
* Remove support for Elasticsearch versions below 6.6.0, add testing for OpenSearch 1.1.0.
* New raw query table function in Elasticsearch can replace old full text search and query pass-through support.

```

**Comment by mosabua** [comment]
@colebow can we get this finished and merged?

**Comment by bitsondatadev** [comment]
Screen Shot 2022-08-04 at 2 35 01 PM
Woah...something is really odd about the broadcast page...I wonder if something isn't formatted correctly in the jekyll header.

**Comment by bitsondatadev** [comment]
> Screen Shot 2022-08-04 at 2 35 01 PM Woah...something is really odd about the broadcast page...I wonder if something isn't formatted correctly in the jekyll header.

Looks like this is resolved.

File: broadcast/index.md [hunk 12]
```diff
@@ -45,7 +45,7 @@ See us live every two weeks!
 <a href="https://podcasts.apple.com/us/podcast/trino-community-broadcast/id1533484786" target="_blank">
   <i class="fab fa-apple watch-listen-icon" title="Apple"></i>
 </a>
-<a href="https://podcasts.google.com/feed/aHR0cHM6Ly9mZWVkcy5idXp6c3Byb3V0LmNvbS8xMzc0NTMyLnJzcw==" target="_blank">
+<a href="https://podcasts.google.com/feed/aHR0cHM6Ly9mYXN0Lndpc3RpYS5jb20vY2hhbm5lbHMvN2Z1dzA0czd2Yy9yc3M" target="_blank">
```
Comment (mosabua) [inline broadcast/index.md:48]: why this change?

File: _episodes/38.md [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+---
+
+## Guests
+
+In this episode we have the pleasure to chat with a couple familiar faces who
+have been hard at work building and understanding the features we're talking
+about today:
+
+* [Kasia Findeisen](https://github.com/kasiafi), Trino Maintainer
+* [Martin Traverso](https://twitter.com/mtraverso), Trino Co-Creator & Maintainer
```
Comment (mosabua) [inline _episodes/38.md:36]:
```suggestion
* [Martin Traverso](https://twitter.com/mtraverso), Trino Cocreator and Maintainer
```

Comment (mosabua) [reply _episodes/38.md:36]: also could go with lowercase..

**Review by mosabua — CHANGES_REQUESTED** [review]
Two minor adjustments but then good to go

File: broadcast/index.md (hunk 12 above)
Comment (colebow) [reply broadcast/index.md:48]: [The previous link](https://podcasts.google.com/feed/aHR0cHM6Ly9mZWVkcy5idXp6c3Byb3V0LmNvbS8xMzc0NTMyLnJzcw==) doesn't work. 

[The new link](https://podcasts.google.com/feed/aHR0cHM6Ly9mYXN0Lndpc3RpYS5jb20vY2hhbm5lbHMvN2Z1dzA0czd2Yy9yc3M) does. I put it in a different commit for clarity's sake.

**Review by bitsondatadev — APPROVED** [review]
Just this nit and that's it!

File: _episodes/38.md [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+
+#### CSVreader
+
+If you've ever tried to create a table from a CSV file, you know it can be a
+painful experience. It has to be very explicit, very diligent, and there's a lot
+of manual cross-checking involved in ensuring that each column aligns perfectly
+and is correctly typed for the columns present in the CSV. Enter polymorphic
+table functions, here to save the day.
+
+The table function itself will be defined on the connector, so all the end user
```
Comment (martint) [inline _episodes/38.md:226]: This seems to imply that 1) we're going to implement this function and 2) it will look exactly as described below. Both of those statements are incorrect. We may never implement this particular function, and if we do, it will likely end up looking very differently.

We should make it clear that these are just some hypothetical examples.

Comment (colebow) [reply _episodes/38.md:226]: I'll update to make it clearer, but we do have a paragraph above (line 170) being very explicit that everything which follows is hypothetical.

**Comment by bitsondatadev** [comment]
@martint, is this good to merge?

**Comment by mosabua** [comment]
Can we please get that merged @kasiafi and @martint ?

**Comment by mosabua** [comment]
Thank you!
